src/model/write/request.rs

```diff
@@ -0,0 +1,582 @@
+        }
+    }
+
+    fn insert(&mut self, name: String) -> u32 {
+        *self.dict.entry(name).or_insert_with(|| {
+            let old_name_idx = self.name_idx;
+            self.name_idx += 1;
+            old_name_idx
+        })
+    }
+
+    fn convert_ordered(self) -> Vec<String> {
+        let mut ordered = vec![String::new(); self.dict.len()];
+        self.dict
+            .into_iter()
+            .for_each(|(name, idx)| ordered[idx as usize] = name);
+        ordered
+    }
+}
+
+impl From<WriteRequest> for WriteRequestPb {
+    fn from(req: WriteRequest) -> Self {
+        let mut req_pb = WriteRequestPb::default();
+        // partition the write entries first
+        let mut partitions_by_metric: HashMap<_, Vec<_>> = HashMap::new();
+        for (idx, entry) in req.write_entries.iter().enumerate() {
+            let parition = partitions_by_metric
+                .entry(entry.series.metric.clone())
+                .or_insert(Vec::new());
+            parition.push(idx);
+        }
+
+        let mut write_metrics = Vec::with_capacity(partitions_by_metric.len());
+        for partition in partitions_by_metric {
+            write_metrics.push(convert_one_write_metric(
+                partition.0,
+                &partition.1,
+                &req.write_entries,
+            ));
+        }
+        req_pb.set_metrics(write_metrics.into());
+
+        req_pb
+    }
+}
+
+fn convert_one_write_metric(
+    metric: String,
+    idxs: &[usize],
+    entries: &[WriteEntry],
+) -> WriteMetricPb {
+    let mut write_metric_pb = WriteMetricPb::default();
+
+    let mut tags_dic = NameDict::new();
+    let mut fields_dic = NameDict::new();
+    let mut wirte_entries = Vec::with_capacity(idxs.len());
+    for idx in idxs {
+        assert!(*idx < entries.len());
+        wirte_entries.push(convert_entry(
+            &mut tags_dic,
+            &mut fields_dic,
+            &entries[*idx],
+        ));
+    }
+
+    write_metric_pb.set_metric(metric);
+    write_metric_pb.set_tag_names(tags_dic.convert_ordered().into());
+    write_metric_pb.set_field_names(fields_dic.convert_ordered().into());
+    write_metric_pb.set_entries(wirte_entries.into());
+
+    write_metric_pb
+}
+
+fn convert_entry(
+    tags_dic: &mut NameDict,
+    fields_dic: &mut NameDict,
+    entry: &WriteEntry,
+) -> WriteEntryPb {
+    let mut entry_pb = WriteEntryPb::default();
+    entry_pb.set_tags(convert_tags(tags_dic, &entry.series.tags).into());
+    entry_pb.set_field_groups(convert_ts_fields(fields_dic, &entry.ts_fields).into());
+
+    entry_pb
+}
+
+// TODO(kamille) reduce cloning from tags.
+fn convert_tags(tags_dic: &mut NameDict, tags: &BTreeMap<String, Value>) -> Vec<TagPb> {
+    if tags.is_empty() {
+        return Vec::new();
+    }
+
+    let mut tag_pbs = Vec::with_capacity(tags.len());
+    for (name, val) in tags {
+        let mut tag_pb = TagPb::default();
+        tag_pb.set_name_index(tags_dic.insert(name.clone()));
```

Comment:
Maybe we should avoid this `name.clone()`.

Reply:
Done